luksmeta.8
'\" t
.\"     Title: luksmeta
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 07/10/2018
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "LUKSMETA" "8" "07/10/2018" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
luksmeta \- Utility for storing metadata in a LUKSv1 header
.SH "SYNOPSIS"
.sp
\fBluksmeta test\fR \-d DEVICE
.sp
\fBluksmeta nuke\fR \-d DEVICE [\-f]
.sp
\fBluksmeta init\fR \-d DEVICE [\-f] [\-n]
.sp
\fBluksmeta show\fR \-d DEVICE [\-s SLOT]
.sp
\fBluksmeta save\fR \-d DEVICE [\-s SLOT] \-u UUID < DATA
.sp
\fBluksmeta load\fR \-d DEVICE \-s SLOT [\-u UUID] > DATA
.sp
\fBluksmeta wipe\fR \-d DEVICE \-s SLOT [\-u UUID] [\-f]
.SH "OVERVIEW"
.sp
The \fBluksmeta\fR utility enables an administrator to store metadata in the gap between the end of the LUKSv1 header and the start of the encrypted data\&. This is useful for storing data that is available before the volume is unlocked, usually for use during the volume unlock process\&.
.sp
The metadata is stored in a series of UUID\-typed slots, allowing multiple blocks of metadata\&. Although the \fBluksmeta\fR slots are inspired by the LUKS slots, they are functionally independent and share only a casual relationship\&. Slots merely provide a hint that a given chunk of metadata is associated with a specific LUKSv1 password (in a slot with the same number)\&. However, \fBluksmeta\fR itself is indifferent to the relationship between a LUKSv1 slot and the correspondly numbered \fBluksmeta\fR slot, with one exception (detailed below)\&.
.sp
After a LUKSv1 volume is initialized using \fBcryptsetup\fR(8), it must also be initialized for metadata storage by \fBluksmeta init\fR\&. Once this is complete, the device is ready to store medata\&.
.sp
Data can be written to a slot using \fBluksmeta save\fR or read from a slot using \fBluksmeta load\fR\&. You can also erase the data in an existing slot using \fBluksmeta wipe\fR or query the slots using \fBluksmeta show\fR\&.
.SH "UUID GENERATION"
.sp
It is often presumed that saving metadata to a slot requires a specific UUID or that there is an official registry of UUID types\&. This is incorrect\&.
.sp
UUID stands for Universally Unique IDentifier\&. UUIDs are a standardized, widely\-used data type used for identification without a central registry\&. For the relevant standards, see ISO 9834\-8:2005 and RFC 4122\&.
.sp
UUIDs are large enough that collision is practically impossible\&. So if your application wants to store data in a \fBluksmeta\fR slot, just generate your own UUID and use it consistently to refer to your type of data\&. If you have multiple types of data, feel free to generate multiple UUIDs\&.
.sp
The easiest way to generate a UUID is to use \fBuuidgen\fR(1)\&. However, any compliant UUID generator will suffice\&.
.SH "INITIALIZATION"
.sp
Before reading or writing metadata, the LUKSv1 block device must be initialized for metadata storage\&. Three commands help with this process: \fBluksmeta test\fR, \fBluksmeta nuke\fR and \fBluksmeta init\fR\&.
.sp
The \fBluksmeta test\fR command simply checks an existing block device to see if it is initialized for metadata storage\&. This command does not provide any output, so be sure to check its return code (see below)\&.
.sp
The \fBluksmeta nuke\fR command will zero (erase) the entire LUKSv1 header gap\&. Since this operation is destructive, user confirmation will be required before clearing the gap unless the \fB\-f\fR option is supplied\&.
.sp
The \fBluksmeta init\fR command initializes the LUKSv1 block device for metadata storage\&. This process will wipe out any data in the LUKSv1 header gap\&. For this reason, this command will require user confirmation before any data is written unless the \fB\-f\fR option is supplied\&. Note that this command succeeds without any modification if the device is already initialized\&. If you would like to force the creation of clean initialization state, you can specify the \fB\-n\fR option to nuke the LUKSv1 header gap before initialization (but after user confirmation)\&.
.SH "METADATA STATE"
.sp
The \fBluksmeta show\fR command displays the current state of slots on the LUKSv1 block device\&. If no slot is specified, it prints a table consisting of the slot number, the corresponding LUKSv1 slot state and the UUID of the data stored in the \fBluksmeta\fR slot (or "empty" if no data is stored)\&. If a slot is specified, this command simply prints out the UUID of the data in the slot\&. If the slot does not contain data, it prints nothing\&.
.SH "MANAGING METADATA"
.sp
Managing the metadata in the slots is performed with three commands: \fBluksmeta save\fR, \fBluksmeta load\fR and \fBluksmeta wipe\fR\&. These commands write metadata to a slot, read metadata from a slot and erase metadata in a slot, respectively\&.
.sp
The \fBluksmeta save\fR command reads metadata on standard input and writes it to the specified slot using the specified UUID\&. If no slot is specified, \fBluksmeta\fR will search for the first slot number for which the LUKSv1 slot is inactive and the \fBluksmeta\fR slot is empty\&. This represents the only official correlation between LUKSv1 slots and \fBluksmeta\fR slots\&. In this case, the metadata is written to the first applicable slot using the specified UUID and the slot number is printed to standard output\&. In either case, this command will never overwrite existing data\&. To replace data in a slot you will need to execute \fBluksmeta wipe\fR before \fBluksmeta save\fR\&.
.sp
The \fBluksmeta load\fR command reads data from the specified slot and writes it to standard output\&. If a UUID is specified, the command will verify that the UUID associated with the metadata in the slot matches the specified UUID\&. This type check helps to ensure that you always receive the type of data you are expecting as output\&. If the UUIDs do not match, the command will fail\&.
.sp
The \fBluksmeta wipe\fR command erases the data from the given slot\&. If a UUID is specified, the command will verify that the UUID associated with the metadata in the slot matches the specified UUID\&. This type check helps to ensure that you only erase the data you intended to erase\&. Because this is a destructive operation, this command will require user confirmation before any data is erased, unless the \fB\-f\fR option is supplied\&. Note that this command succeeds if you attempt to wipe a slot that is already empty\&.
.SH "CAVEATS"
.sp
The amount of storage in the LUKSv1 header gap is extremely limited\&. It also varies based upon the configuration used by LUKSv1 at device initialization time\&. In some LUKSv1 configurations, there is not even enough space for all the metadata slots even at the smallest possible slot size\&.
.sp
During the design of this utility, we considered it likely that users would want to reduce the number of usable slots in exchange for more storage space in the slots used\&. In order to provide this flexibility, the amount of storage available per\-slot is dynamic\&. Put simply, slots are not a fixed size\&. This means that it is possible (and even somewhat likely) to encounter an error during \fBluksmeta save\fR indicating that there is insufficient space\&.
.sp
This error is not a programming bug\&. If you encounter this error it likely means that either all space is being consumed by the already\-written slots or that the metadata you are attempting to write simply does not fit\&.
.sp
You can attempt to resolve this problem by calling \fBluksmeta wipe\fR on slots that are no longer in use\&. This will release the storage space for use by other slots\&. Note that \fBluksmeta\fR does not, however, currently perform defragmentation since the number of usable blocks is rather limited\&. You can attempt to manually get around this by extracting all slot data, wiping the slots and reloading them in order\&. However, this operation is potentially dangerous and should be undertaken with great care\&.
.SH "OPTIONS"
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-d\fR
\fIDEVICE\fR,
\fB\-\-device\fR=\fIDEVICE\fR
: The device on which to perform the operation\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-s\fR
\fISLOT\fR,
\fB\-\-slot\fR=\fISLOT\fR
: The slot number on which to perform the operation\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-u\fR
\fIUUID\fR,
\fB\-\-uuid\fR=\fIUUID\fR
: The UUID to associate with the operation\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fB\-f\fR,
\fB\-\-force\fR
: Forcibly suppress all user prompting\&.
.RE
.SH "RETURN VALUES"
.sp
This command uses the return values as defined by \fBsysexit\&.h\fR\&. The following are general errors whose meaning is shared by all \fBluksmeta\fR commands:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBEX_OK\fR
: The operation was successful\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBEX_OSERR\fR
: An undefined operating system error occurred\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBEX_USAGE\fR
: The program was called with invalid parameters\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBEX_IOERR\fR
: An IO error occurred when writing to the device\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBEX_OSFILE\fR
: The device is not initialized or is corrupted\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBEX_NOPERM\fR
: The user did not grant permission during confirmation\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBEX_NOINPUT\fR
: An error occurred while reading from standard input\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBEX_DATAERR\fR
: The specified UUID does not match the slot UUID\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fBEX_CANTCREAT\fR
: There is insufficient space in LUKSv1 header\&.
.RE
.sp
Additionally, \fBluksmeta save\fR will return \fBEX_UNAVAILABLE\fR when you attempt to save data into a slot that is already used\&. Likewise, \fBluksmeta load\fR will return \fBEX_UNAVAILABLE\fR when you attempt to read from an empty slot\&.
.SH "EXAMPLES"
.sp
Destroy all data (including LUKSMeta data) in the LUKSv1 header gap and initialize the gap for LUKSMeta storage:
.sp
.if n \{\
.RS 4
.\}
.nf
$ luksmeta init \-n \-f \-d /dev/sdz
.fi
.if n \{\
.RE
.\}
.sp
If already initialized, do nothing\&. Otherwise, destroy all non\-LUKSMeta data in the LUKSv1 header gap and initialize the gap for LUKSMeta storage:
.sp
.if n \{\
.RS 4
.\}
.nf
$ luksmeta init \-f \-d /dev/sdz
.fi
.if n \{\
.RE
.\}
.sp
Write some data to a slot:
.sp
.if n \{\
.RS 4
.\}
.nf
$ UUID=*uuidgen*
$ echo $UUID
31c25e3b\-b8e2\-4eaa\-a427\-23aa882feef2
$ echo "Hello, World" | luksmeta save \-d /dev/sdz \-s 0 \-u $UUID
.fi
.if n \{\
.RE
.\}
.sp
Read the data back:
.sp
.if n \{\
.RS 4
.\}
.nf
$ luksmeta load \-d /dev/sdz \-s 0 \-u $UUID
Hello, World
.fi
.if n \{\
.RE
.\}
.sp
Wipe the data from the slot:
.sp
.if n \{\
.RS 4
.\}
.nf
$ luksmeta wipe \-d /dev/sdz \-s 0 \-u $UUID
.fi
.if n \{\
.RE
.\}
.sp
Erase all trace of LUKSMeta:
.sp
.if n \{\
.RS 4
.\}
.nf
$ luksmeta nuke \-f \-d /dev/sdz
.fi
.if n \{\
.RE
.\}
.SH "AUTHOR"
.sp
Nathaniel McCallum <npmccallum@redhat\&.com>
.SH "SEE ALSO"
.sp
\fBcryptsetup\fR(8), \fBuuidgen\fR(1)


luksmeta.c
/* vim: set tabstop=8 shiftwidth=4 softtabstop=4 expandtab smarttab colorcolumn=80: */
/*
 * Copyright (c) 2016 Red Hat, Inc.
 * Author: Nathaniel McCallum <npmccallum@redhat.com>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "luksmeta.h"

#include <errno.h>
#include <getopt.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sysexits.h>

#define UUID_TMPL \
    "%02hhx%02hhx%02hhx%02hhx-" \
    "%02hhx%02hhx-%02hhx%02hhx-%02hhx%02hhx-" \
    "%02hhx%02hhx%02hhx%02hhx%02hhx%02hhx"

#define UUID_ARGS(u) \
    u[0x0], u[0x1], u[0x2], u[0x3], u[0x4], u[0x5], u[0x6], u[0x7], \
    u[0x8], u[0x9], u[0xa], u[0xb], u[0xc], u[0xd], u[0xe], u[0xf]

struct options {
    const char *device;
    luksmeta_uuid_t uuid;
    bool have_uuid;
    bool force;
    bool nuke;
    int slot;
};

static int
cmd_test(const struct options *opts, struct crypt_device *cd)
{
    return luksmeta_test(cd) == 0 ? EX_OK : EX_OSFILE;
}

static int
cmd_nuke(const struct options *opts, struct crypt_device *cd)
{
    int r = 0;

    if (!opts->force) {
        int c = 'X';

        fprintf(stderr,
                "You are about to erase all data in the LUKSMeta storage area.\n"
                "A backup is advised before erasure is performed.\n\n");

        while (!strchr("YyNn", c)) {
            fprintf(stderr, "Do you wish to nuke %s? [yn] ",
                    crypt_get_device_name(cd));
            c = getc(stdin);
        }

        if (strchr("Nn", c))
            return EX_NOPERM;
    }

    r = luksmeta_nuke(cd);
    switch (r) {
    case 0:
        return EX_OK;

    default:
        fprintf(stderr, "Error while erasing device (%s): %s\n",
                opts->device, strerror(-r));
        return EX_OSERR;
    }
}

static int
cmd_init(const struct options *opts, struct crypt_device *cd)
{
    int r = 0;

    if (!opts->force) {
        int c = 'X';

        fprintf(stderr,
            "You are about to initialize a LUKS device for metadata storage.\n"
            "Attempting to initialize it may result in data loss if data was\n"
            "already written into the LUKS header gap in a different format.\n"
            "A backup is advised before initialization is performed.\n\n");

        while (!strchr("YyNn", c)) {
            fprintf(stderr, "Do you wish to initialize %s? [yn] ",
                    crypt_get_device_name(cd));
            c = getc(stdin);
        }

        if (strchr("Nn", c))
            return EX_NOPERM;
    }

    if (opts->nuke) {
        struct options o = { .force = true, .device = opts->device };
        r = cmd_nuke(&o, cd);
        if (r != EX_OK)
            return r;
    }

    r = luksmeta_init(cd);
    switch (r) {
    case 0: /* fallthrough */
    case -EALREADY:
        return EX_OK;

    case -ENOSPC:
        fprintf(stderr, "Insufficient space in the LUKS header (%s)\n",
                opts->device);
        return EX_CANTCREAT;

    default:
        fprintf(stderr, "Error while initializing device (%s): %s\n",
                opts->device, strerror(-r));
        return EX_OSERR;
    }

    return EX_OK;
}

static const char *
status(struct crypt_device *cd, int keyslot)
{
    switch (crypt_keyslot_status(cd, keyslot)) {
    case CRYPT_SLOT_INVALID: return "invalid";
    case CRYPT_SLOT_INACTIVE: return "inactive";
    case CRYPT_SLOT_ACTIVE: return "active";
    case CRYPT_SLOT_ACTIVE_LAST: return "active";
    default: return "unknown";
    }
}

static int
cmd_show(const struct options *opts, struct crypt_device *cd)
{
    luksmeta_uuid_t uuid = {};

    for (int i = 0, r = 0; i < crypt_keyslot_max(CRYPT_LUKS1); i++) {
        if (opts->slot >= 0 && i != opts->slot)
            continue;

        r = luksmeta_load(cd, i, uuid, NULL, 0);
        switch (r) {
        case -EBADSLT:
            fprintf(stderr, "Invalid slot (%d)\n", opts->slot);
            return EX_USAGE;

        case -ENOENT:
            fprintf(stderr, "Device is not initialized (%s)\n",
                    opts->device);
            return EX_OSFILE;

        case -EINVAL:
            fprintf(stderr, "LUKSMeta data appears corrupt (%s)\n",
                    opts->device);
            return EX_OSFILE;

        case -ENODATA:
            if (opts->slot < 0)
                fprintf(stdout, "%d %8s %s\n", i, status(cd, i), "empty");
            break;

        default:
            if (r < 0) {
                fprintf(stderr, "%d %8s %s\n",
                        i, status(cd, i), "unknown error");
            } else {
                if (opts->slot < 0)
                    fprintf(stdout, "%d %8s ", i, status(cd, i));

                fprintf(stdout, UUID_TMPL "\n", UUID_ARGS(uuid));
            }
            break;
        }
    }

    return EX_OK;
}

static int
cmd_save(const struct options *opts, struct crypt_device *cd)
{
    uint8_t *in = NULL;
    size_t inl = 0;
    int r = 0;

    if (!opts->have_uuid) {
        fprintf(stderr, "UUID required\n");
        return EX_USAGE;
    }

    while (!feof(stdin)) {
        uint8_t *tmp = NULL;

        tmp = realloc(in, inl + 4096);
        if (!tmp) {
            fprintf(stderr, "Out of memory\n");
            free(in);
            return EX_OSERR;
        }

        in = tmp;
        r = fread(&in[inl], 1, 4096, stdin);
        inl += r;
        if (r < 4096 && (ferror(stdin) || inl == 0)) {
            fprintf(stderr, "Error reading from standard input\n");
            free(in);
            return EX_NOINPUT;
        }
    }

    if (!in) {
        fprintf(stderr, "No data on standard input\n");
        return EX_NOINPUT;
    }

    r = luksmeta_save(cd, opts->slot, opts->uuid, in, inl);
    memset(in, 0, inl);
    free(in);
    switch (r) {
    case -ENOENT:
        fprintf(stderr, "Device is not initialized (%s)\n", opts->device);
        return EX_OSFILE;

    case -EINVAL:
        fprintf(stderr, "LUKSMeta data appears corrupt (%s)\n", opts->device);
        return EX_OSFILE;

    case -EBADSLT:
        fprintf(stderr, "The specified slot is invalid (%d)\n", opts->slot);
        return EX_USAGE;

    case -EKEYREJECTED:
        fprintf(stderr, "The specified UUID is reserved (" UUID_TMPL ")\n",
                UUID_ARGS(opts->uuid));
        return EX_USAGE;

    case -EALREADY:
        fprintf(stderr, "Will not overwrite existing slot (%d)\n", opts->slot);
        return EX_UNAVAILABLE;

    case -ENOSPC:
        fprintf(stderr, "Insufficient space in the LUKS header (%s)\n",
                opts->device);
        return EX_CANTCREAT;

    default:
        if (r < 0)
            fprintf(stderr, "An unknown error occurred\n");
        else if (opts->slot < 0)
            fprintf(stdout, "%d\n", r);

        return r < 0 ? EX_OSERR : EX_OK;
    }
}

static int
cmd_load(const struct options *opts, struct crypt_device *cd)
{
    luksmeta_uuid_t uuid = {};
    int r = 0;

    if (opts->slot < 0) {
        fprintf(stderr, "Slot required\n");
        return EX_USAGE;
    }

    r = luksmeta_load(cd, opts->slot, uuid, NULL, 0);
    if (r >= 0) {
        uint8_t *out = NULL;

        if (opts->have_uuid && memcmp(opts->uuid, uuid, sizeof(uuid)) != 0) {
            fprintf(stderr,
                    "The given UUID does not match the slot UUID:\n"
                    "UUID: " UUID_TMPL "\n"
                    "SLOT: " UUID_TMPL "\n",
                    UUID_ARGS(opts->uuid),
                    UUID_ARGS(uuid));
            return EX_DATAERR;
        }

        out = malloc(r);
        if (!out) {
            fprintf(stderr, "Out of memory!\n");
            return EX_OSERR;
        }

        r = luksmeta_load(cd, opts->slot, uuid, out, r);
        if (r >= 0) {
            fwrite(out, 1, r, stdout);
            memset(out, 0, r);
        }

        free(out);
    }

    switch (r) {
    case -ENOENT:
        fprintf(stderr, "Device is not initialized (%s)\n", opts->device);
        return EX_OSFILE;

    case -EINVAL:
        fprintf(stderr, "LUKSMeta data appears corrupt (%s)\n", opts->device);
        return EX_OSFILE;

    case -EBADSLT:
        fprintf(stderr, "The specified slot is invalid (%d)\n", opts->slot);
        return EX_USAGE;

    case -ENODATA:
        fprintf(stderr, "The specified slot is empty (%d)\n", opts->slot);
        return EX_UNAVAILABLE;

    default:
        if (r < 0)
            fprintf(stderr, "An unknown error occurred\n");

        return r < 0 ? EX_OSERR : EX_OK;
    }
}

static int
cmd_wipe(const struct options *opts, struct crypt_device *cd)
{
    luksmeta_uuid_t uuid = {};
    int r = 0;

    if (opts->slot < 0) {
        fprintf(stderr, "Slot required\n");
        return EX_USAGE;
    }

    if (!opts->force) {
        int c = 'X';

        fprintf(stderr,
            "You are about to wipe a slot. This operation is unrecoverable.\n"
            "A backup is advised before proceeding.\n\n");

        while (!strchr("YyNn", c)) {
            fprintf(stderr, "Do you wish to erase slot %d on %s? [yn] ",
                    opts->slot, crypt_get_device_name(cd));
            c = getc(stdin);
        }

        if (strchr("Nn", c))
            return EX_NOPERM;
    }

    r = luksmeta_wipe(cd, opts->slot, opts->have_uuid ? opts->uuid : NULL);
    switch (r) {
    case -EALREADY:
        return EX_OK;

    case -ENOENT:
        fprintf(stderr, "Device is not initialized (%s)\n", opts->device);
        return EX_OSFILE;

    case -EINVAL:
        fprintf(stderr, "LUKSMeta data appears corrupt (%s)\n", opts->device);
        return EX_OSFILE;

    case -EBADSLT:
        fprintf(stderr, "The specified slot is invalid (%d)\n", opts->slot);
        return EX_USAGE;

    case -EKEYREJECTED:
        r = luksmeta_load(cd, opts->slot, uuid, NULL, 0);
        if (r >= 0) {
            fprintf(stderr,
                    "The given UUID does not match the slot UUID:\n"
                    "UUID: " UUID_TMPL "\n"
                    "SLOT: " UUID_TMPL "\n",
                    UUID_ARGS(opts->uuid),
                    UUID_ARGS(uuid));
        } else {
            fprintf(stderr,
                    "The given UUID does not match the slot UUID:\n"
                    "UUID: " UUID_TMPL "\n"
                    "SLOT: UNKNOWN\n",
                    UUID_ARGS(opts->uuid));
        }
        return EX_DATAERR;

    default:
        if (r < 0)
            fprintf(stderr, "An unknown error occurred\n");

        return r < 0 ? EX_OSERR : EX_OK;
    }
}

static const char *sopts = "hfnd:u:s:";
static const struct option lopts[] = {
    { "help",                      .val = 'h' },
    { "nuke",   no_argument,       .val = 'n' },
    { "force",  no_argument,       .val = 'f' },
    { "device", required_argument, .val = 'd' },
    { "uuid",   required_argument, .val = 'u' },
    { "slot",   required_argument, .val = 's' },
    {}
};

static const struct {
    int (*func)(const struct options *opts, struct crypt_device *cd);
    const char *name;
} commands[] = {
    { cmd_test, "test", },
    { cmd_nuke, "nuke", },
    { cmd_init, "init", },
    { cmd_show, "show", },
    { cmd_save, "save", },
    { cmd_load, "load", },
    { cmd_wipe, "wipe", },
    {}
};

int
main(int argc, char *argv[])
{
    struct options o = { .slot = CRYPT_ANY_SLOT };

    for (int c; (c = getopt_long(argc, argv, sopts, lopts, NULL)) != -1; ) {
        switch (c) {
        case 'h': goto usage;
        case 'd': o.device = optarg; break;
        case 'n': o.nuke = true; break;
        case 'f': o.force = true; break;
        case 'u':
            if (sscanf(optarg, UUID_TMPL, UUID_ARGS(&o.uuid)) != 16) {
                fprintf(stderr, "Invalid UUID (%s)\n", optarg);
                return EX_USAGE;
            }

            o.have_uuid = true;
            break;
        case 's':
            if (sscanf(optarg, "%d", &o.slot) != 1 || o.slot < 0 ||
                o.slot >= crypt_keyslot_max(CRYPT_LUKS1)) {
                fprintf(stderr, "Invalid slot (%s)\n", optarg);
                return EX_USAGE;
            }
            break;
        }
    }

    if (argc > 1 && !o.device) {
        fprintf(stderr, "Device must be specified\n\n");
        goto usage;
    }

    if (optind != argc - 1)
        goto usage;

    for (size_t i = 0; argc > 1 && commands[i].name; i++) {
        struct crypt_device *cd = NULL;
        const char *type = NULL;
        int r = 0;

        if (strcmp(argv[optind], commands[i].name) != 0)
            continue;

        r = crypt_init(&cd, o.device);
        if (r != 0) {
            fprintf(stderr, "Unable to open device (%s): %s\n",
                    o.device, strerror(-r));
            return EX_IOERR;
        }

        r = crypt_load(cd, NULL, NULL);
        if (r != 0) {
            fprintf(stderr, "Unable to read LUKSv1 header (%s): %s\n",
                    o.device, strerror(-r));
            crypt_free(cd);
            return EX_IOERR;
        }

        type = crypt_get_type(cd);
        if (type == NULL) {
            fprintf(stderr, "Unable to determine device type for %s\n",
                    o.device);
            crypt_free(cd);
            return EX_OSFILE;
        }

        if (strcmp(type, CRYPT_LUKS1) != 0) {
            fprintf(stderr, "%s (%s) is not a LUKSv1 device\n", o.device, type);
            crypt_free(cd);
            return EX_OSFILE;
        }

        r = commands[i].func(&o, cd);
        crypt_free(cd);
        return r;
    }

    fprintf(stderr, "Invalid command\n\n");

usage:
    fprintf(stderr,
            "Usage: luksmeta test -d DEVICE\n"
            "   or: luksmeta nuke -d DEVICE [-f]\n"
            "   or: luksmeta init -d DEVICE [-f] [-n]\n"
            "   or: luksmeta show -d DEVICE [-s SLOT]\n"
            "   or: luksmeta save -d DEVICE [-s SLOT]  -u UUID  < DATA\n"
            "   or: luksmeta load -d DEVICE  -s SLOT  [-u UUID] > DATA\n"
            "   or: luksmeta wipe -d DEVICE  -s SLOT  [-u UUID] [-f]\n");
    return EX_USAGE;
}
